src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs

```diff
@@ -760,73 +776,171 @@ private DictionarySpec CreateDictionarySpec(INamedTypeSymbol type, Location? loc
-                spec.ConcreteType = ConstructGenericCollectionTypeSpec(concreteType, elementType);
-                spec.PopulationCastType = ConstructGenericCollectionTypeSpec(populationCastType, elementType);
+                spec.ConcreteType = ConstructGenericCollectionSpecIfRequired(concreteType, elementType);
+                spec.PopulationCastType = ConstructGenericCollectionSpecIfRequired(populationCastType, elementType);
 
                 return spec;
             }
 
             private ObjectSpec? CreateObjectSpec(INamedTypeSymbol type, Location? location)
             {
-                Debug.Assert(!_createdSpecs.ContainsKey(type));
-
                 // Add spec to cache before traversing properties to avoid stack overflow.
-                if (!HasPublicParameterlessCtor(type))
+                ObjectSpec objectSpec = new(type) { Location = location };
+                _createdSpecs.Add(type, objectSpec);
+
+                string typeName = objectSpec.Name;
+                IMethodSymbol? ctor = null;
+                DiagnosticDescriptor? diagnosticDescriptor = null;
+
+                if (!(type.IsAbstract || type.TypeKind is TypeKind.Interface))
                 {
-                    ReportUnsupportedType(type, ParserDiagnostics.NeedPublicParameterlessConstructor, location);
-                    _createdSpecs.Add(type, null);
-                    return null;
+                    IMethodSymbol? parameterlessCtor = null;
+                    IMethodSymbol? parameterizedCtor = null;
+                    bool hasMultipleParameterizedCtors = false;
+
+                    foreach (IMethodSymbol candidate in type.InstanceConstructors)
+                    {
+                        if (candidate.DeclaredAccessibility is not Accessibility.Public)
+                        {
+                            continue;
+                        }
+
+                        if (candidate.Parameters.Length is 0)
+                        {
+                            parameterlessCtor = candidate;
+                        }
+                        else if (parameterizedCtor is not null)
+                        {
+                            hasMultipleParameterizedCtors = true;
+                        }
+                        else
+                        {
+                            parameterizedCtor = candidate;
+                        }
+                    }
+
+                    bool hasPublicParameterlessCtor = type.IsValueType || parameterlessCtor is not null;
+                    if (!hasPublicParameterlessCtor && hasMultipleParameterizedCtors)
+                    {
+                        diagnosticDescriptor = ParserDiagnostics.MultipleParameterizedConstructors;
+                        objectSpec.InitExceptionMessage = string.Format(ExceptionMessages.MultipleParameterizedConstructors, typeName);
+                    }
+
+                    ctor = parameterizedCtor ?? parameterlessCtor;
```

Comment:
>ctor = parameterizedCtor ?? parameterlessCtor; [](http://example.com/codeflow?start=20&length=46)

does this match the runtime behavior? don't we prefer parameter-less constructor over the parameterized ones?